### PR TITLE
[SPARK-35774][SQL] Parse any year-month interval types in SQL

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR extends the parser rules to be able to parse the following types:

* INTERVAL YEAR
* INTERVAL YEAR TO MONTH
* INTERVAL MONTH

### Why are the changes needed?

For ANSI compliance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New assertion.

### CODE REVIEW

**Comment by sarutak** [comment]
cc: @MaxGekk

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @sarutak .

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -2514,8 +2514,13 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
   }
 
   override def visitYearMonthIntervalDataType(ctx: YearMonthIntervalDataTypeContext): DataType = {
-    // TODO(SPARK-35774): Parse any year-month interval types in SQL
-    YearMonthIntervalType()
+    val start = YearMonthIntervalType.stringToField(ctx.from.getText.toLowerCase(Locale.ROOT))
+    val end = if (ctx.to != null) {
+      YearMonthIntervalType.stringToField(ctx.to.getText.toLowerCase(Locale.ROOT))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2519]: shall we forbid `interval year to year`?

Comment (sarutak) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2519]: Sounds reasonable. I'll open a following-up PR for year-month/day-time intervals.

**Pull request closed without merging**